Clarify that browsers are not the target implementation for the W3C TraceContext specification (#50)
w3c/distributed-tracing-wg: EXPLAINER.md

# Distributed Tracing Working Group

## Introduction
The Distributed Tracing Working Group currently drives the standardization of two specifications:

- W3C TraceContext

> - W3C Baggage

Reviewer: Suggestion: link the above items to the respective specs and delete everything below.

Author: Sure, I will add links to the respective specs, however my understanding is that this "explainer" file/content is still needed during various W3C reviews (https://tag.w3.org/explainers/). For example, I have seen various reviews asking for a link to the explainer file. Would be happy to update this content though (to make it crisper or to elaborate where needed).

Let's dive more into what these specifications cover and the need for them.

## W3C TraceContext

### What is W3C TraceContext?
The W3C TraceContext specification defines a universally agreed upon format to propagate trace context data. It specifies standard HTTP headers and standard value formats. It also defines a mechanism to forward vendor-specific trace data.

> These headers can also be propagated by systems that are not monitored but still need to participate in the trace to allow end-to-end tracing. The target implementation for the current version of the specification is applications and services; browsers/user-agents are not in scope.

Reviewer 2: IMO it might be good to briefly talk about the target audience in the first subsection for trace context. Something like this maybe?

"The intended audience for this specification are observability solutions, that is, tracing platforms or application performance monitoring providers."

![Trace Context Propagation](./assets/explainer_all_in_one.png "Trace Context Propagation")

### What problem does it solve: Improving the Observability of complex distributed systems
Understanding what is happening in a complex distributed system can be a challenge.

In a complex distributed system, a single logical operation can involve work done by hundreds of participants. When an operation is delayed or encountered an error in such a system, you want to get to one logical explanation of what is happening to that operation. But the distributed nature of this system makes it a challenge. Sure, each participant will have its own view of how it handled its request in the context of the overall operation.

But how do you tie them all together to understand what is happening? More generally, how do we gain the ability to answer questions about a system to understand it better: who all participated in handling an operation? What is the critical path? Where do failures happen? What characteristics did orders that took more than 20 minutes have in common? What are the performance trends over time? What are the reliability trends over time?

> You should be able to ask any such question and get answers. This specification helps improve the Observability of systems in an open and vendor-neutral way.

Reviewer 2: ```suggestion
You should be able to ask any such question and get answers. This specification helps improve the observability of systems in an open and vendor-neutral way.
```

There are a couple of places where important terms are written capitalized. Is there a style guide that suggests to do that, which I am not aware of? If so, please disregard the suggestions to make them lower case.

### The building blocks needed to tie things together into a "Distributed Trace"
Let’s look at a few key building blocks needed to tie things together:

First, we need a way to *uniquely identify a logical operation* flowing through a system.

> We also *need a way for each participant to know* this unique identifier. This is so that each participant can associate what it is doing to handle its part of that operation. But how can each participant know this? One participant would have to first create this identifier, and then it must be somehow “propagated” to the others. This mechanism must be protocol specific: e.g., for HTTP, this must be through a HTTP header.

Reviewer 2: _suggestion:_

```suggestion
We also *need a way for each participant to know* this unique identifier. This is so that each participant can associate what it is doing to handle its part of that operation. But how can each participant know this? One participant would have to first create this identifier, and then it needs to be propagated all other participants. The propagation mechanism will be protocol specific: e.g., for HTTP, an obvious choice is to use HTTP headers for propagation.
```

> We also then a *need a way for each participant to emit what it did*. We should be able to uniquely identify (within the scope of a single logical operation) the above unit of work done by each participant.

Reviewer 2: > We also then a *need a way for each participant to emit what it did*.

To me this first sentence sounds a bit like it implies that the trace context spec would also deal with _how_ participants emit what they did -- e.g. the trace context spec would talk about _spans_ etc. The only concept we talk about are the span _ids_, though.

Maybe we can omit the first sentence completely and only state the requirement for the span id?

```suggestion
We also need to be able to uniquely identify (within the scope of a single logical operation) the unit of work done by each participant.
```

But the above is not enough: we also need to establish the causality of the work done by different participants. For this, we need a way to capture the parent/child relationships: when a participant calls another, the callee/child needs to know the unique identifier of the caller/parent’s unit of work, so that the child can emit it. How does the child get this? It must be propagated.

If we have all the above, we will gain the ability to connect the dots together.

> The set of related events emitted by various participants, triggered by the same logical operation, is called a Distributed Trace. The context that is propagated between the participants is referred to as the “trace context” that includes the identifiers described above.

Reviewer 2: ```suggestion
The set of related events emitted by various participants, triggered by the same logical operation, is called a distributed trace. The context that is propagated between the participants is referred to as the “trace context” that includes the identifiers described above.
```

### Why does it have to be a standard?
If there is no standardized way to propagate the trace context, each tracing system would have to create its own method. Typically, this would be in the form of a custom HTTP header. If an operation is handled by multiple participants using different tracing systems, such headers will not be propagated between them and the trace will be broken. When this happens, we will not able to reason about the whole system.

To make things interoperable, we need a standard mechanism that everybody can agree to use.

### Goals

> - To provide a standard mechanism to propagate shared context across various applications and services so that multiple tracing platforms can be used in concert with each other.

Reviewer 2: _suggestion:_ We already defined the term "trace context" above, I think we should use it here and not go back to a more generic term like "shared context".

```suggestion
- To provide a standard mechanism to propagate trace context across various applications and services so that multiple tracing platforms can be used in concert with each other.
```

> - To provide a method for application developers to trace requests without invading users' privacy.

Reviewer 2: Is that an actual goal? We spend a lot of time in the most recent privacy review to explain to reviewers that most privacy concerns are orthogonal to what we are doing in the spec.

Author: Since the specification provides guidance on how to generate TraceID and SpanID using mechanisms (generating them using in a pseudo-random way) that respect user's privacy, I think there may still be value in keeping this goal. Please let me know if you disagree.

The traceparent header is a request header which contains a trace id, a parent span id, and trace flags. The traceflags includes information such as if a particular request is sampled. When a tracing system transmits a span to the tracing backend, it includes the Trace ID, the span ID, and the parent span ID (if one exists). The tracing backend uses this information to construct the directed acyclic graph which represents the trace. In a system with a single tracing platform, it is the only header necessary to complete a distributed trace.

###### Tracestate

The tracestate header is a request header which allows a tracing platform to transmit platform-specific information, even through systems that are traced using other tracing platforms. Each tracing platform uses a single key, and the value is treated as an opaque token by other tracing platforms. Each tracing platform is required to forward tracestate keys and values.

### Examples

#### Context loss using a proprietary header

1. Service A calls Service B through an API gateway.
2. The tracing system uses a proprietary header to propagate its trace ID.
3. The API gateway is not configured to propagate the proprietary header.

> 4. Service C does not receive a trace ID and context is lost.

Reviewer 2: I know you only added a `.` at the end here, but wouldn't that need to read 

```suggestion
4. Service B does not receive a trace ID and context is lost.
```

to make sense? The example is A -> API gateway -> B, right?

![Context loss due to a middleware](./assets/explainer_context_loss.png "Context loss due to a middleware")

#### Context propagation with W3C Trace Context
1. Service A calls Service B through an API gateway.
2. The tracing system uses the W3C Trace Context headers.

> 3. The API gateway recognizes the headers as standards-compliant and propagates them.

Reviewer 2: ```suggestion
3. The API gateway recognizes the headers as standard-compliant and propagates them.
```

> 4. Service C receives the trace ID and continues the trace.

Reviewer 2: see above

```suggestion
4. Service B receives the trace ID and continues the trace.
```

![Context propagation through a standard compliant middleware](./assets/explainer_context_preserved.png "Context propagation through a standard compliant middleware")

## W3C Baggage

### What is W3C Baggage?
This specification defines a standard for representing and propagating a set of application-defined properties associated with a distributed request or workflow execution.

> This is independent of the Trace Context specification. Baggage can be used regardless of whether Distributed Tracing is used. This specification standardizes representation and propagation of application-defined properties. In contrast, Trace Context specification standardizes representation and propagation of the metadata needed to enable Distributed Tracing scenarios.

Reviewer 2: ```suggestion
This is independent of the Trace Context specification. Baggage can be used regardless of whether distributed tracing is used. This specification standardizes representation and propagation of application-defined properties. In contrast, Trace Context specification standardizes representation and propagation of the metadata needed to enable distributed tracing scenarios.
```
